File: 2015 ICPC North American Qualifier Contest/undetected.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int x, y;
} Point;

typedef struct {
    int* p;
    int* rank;
    int* setSize;
    int numSets;
} UnionFind;

UnionFind* createUnionFind(int n) {
    UnionFind* uf = (UnionFind*)malloc(sizeof(UnionFind));
    uf->p = (int*)malloc(n * sizeof(int));
    uf->rank = (int*)calloc(n, sizeof(int));
    uf->setSize = (int*)malloc(n * sizeof(int));
    uf->numSets = n;
    
    for (int i = 0; i < n; i++) {
        uf->p[i] = i;
        uf->setSize[i] = 1;
    }
    return uf;
}

int findSet(UnionFind* uf, int i) {
    return (uf->p[i] == i) ? i : (uf->p[i] = findSet(uf, uf->p[i]));
}

int isSameSet(UnionFind* uf, int i, int j) {
    return findSet(uf, i) == findSet(uf, j);
}

void unionSet(UnionFind* uf, int i, int j) {
    if (!isSameSet(uf, i, j)) {
        uf->numSets--;
        int x = findSet(uf, i);
        int y = findSet(uf, j);
        
        if (uf->rank[x] > uf->rank[y]) {
            uf->p[y] = x;
            uf->setSize[x] += uf->setSize[y];
        } else {
            uf->p[x] = y;
            uf->setSize[y] += uf->setSize[x];
            if (uf->rank[x] == uf->rank[y])
                uf->rank[y]++;
        }
    }
}

void freeUnionFind(UnionFind* uf) {
    free(uf->p);
    free(uf->rank);
    free(uf->setSize);
    free(uf);
}

int main() {
    int n;
    scanf("%d", &n);
    
    int* rads = (int*)malloc(n * sizeof(int));
    Point* coords = (Point*)malloc(n * sizeof(Point));
    
    for (int i = 0; i < n; i++) {
        int x, y, r;
        scanf("%d %d %d", &x, &y, &r);
        rads[i] = r;
        coords[i] = (Point){x, y};
    }
    
    UnionFind* uf = createUnionFind(n + 2);
    int i = 0;
    
    while (i < n) {
        // Check left wall collision
        if (coords[i].x - rads[i] < 0)
            unionSet(uf, 0, i + 2);
        // Check right wall collision
        if (coords[i].x + rads[i] > 200)
            unionSet(uf, 1, i + 2);
        
        // Check intersection with previous circles
        for (int j = 0; j < i; j++) {
            int dx = coords[i].x - coords[j].x;
            int dy = coords[i].y - coords[j].y;
            int r_sum = rads[i] + rads[j];
            
            if (dx * dx + dy * dy < r_sum * r_sum)
                unionSet(uf, i + 2, j + 2);
        }
        
        // Check if left and right walls are connected
        if (isSameSet(uf, 0, 1))
            break;
            
        i++;
    }
    
    printf("%d\n", i);
    
    free(rads);
    free(coords);
    freeUnionFind(uf);
    
    return 0;
}

File: 2015 ICPC North American Qualifier Contest/undetected.java
import java.util.*;

class UnionFind {
    private int[] p, rank, setSize;
    public int numSets;
    
    public UnionFind(int n) {
        numSets = n;
        setSize = new int[n];
        rank = new int[n];
        p = new int[n];
        for (int i = 0; i < n; i++) {
            p[i] = i;
            setSize[i] = 1;
        }
    }
    
    public int findSet(int i) {
        return (p[i] == i) ? i : (p[i] = findSet(p[i]));
    }
    
    public boolean isSameSet(int i, int j) {
        return findSet(i) == findSet(j);
    }
    
    public void unionSet(int i, int j) {
        if (!isSameSet(i, j)) {
            numSets--;
            int x = findSet(i);
            int y = findSet(j);
            
            if (rank[x] > rank[y]) {
                p[y] = x;
                setSize[x] += setSize[y];
            } else {
                p[x] = y;
                setSize[y] += setSize[x];
                if (rank[x] == rank[y])
                    rank[y]++;
            }
        }
    }
}

public class undetected {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        
        int[] rads = new int[n];
        int[][] coords = new int[n][2];
        
        for (int i = 0; i < n; i++) {
            coords[i][0] = scanner.nextInt();
            coords[i][1] = scanner.nextInt();
            rads[i] = scanner.nextInt();
        }
        
        UnionFind uf = new UnionFind(n + 2);
        int idx = 0;
        
        while (idx < n) {
            // Check left wall collision
            if (coords[idx][0] - rads[idx] < 0)
                uf.unionSet(0, idx + 2);
            // Check right wall collision
            if (coords[idx][0] + rads[idx] > 200)
                uf.unionSet(1, idx + 2);
            
            // Check intersection with previous circles
            for (int j = 0; j < idx; j++) {
                int dx = coords[idx][0] - coords[j][0];
                int dy = coords[idx][1] - coords[j][1];
                int rSum = rads[idx] + rads[j];
                
                if (dx * dx + dy * dy < rSum * rSum)
                    uf.unionSet(idx + 2, j + 2);
            }
            
            // Check if left and right walls are connected
            if (uf.isSameSet(0, 1))
                break;
                
            idx++;
        }
        
        System.out.println(idx);
        
        scanner.close();
    }
}